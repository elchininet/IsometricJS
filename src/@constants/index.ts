export const SVG_NAMESPACE = 'http://www.w3.org/2000/svg';

export enum SVG_ELEMENTS {
    svg = 'svg',
    group = 'group',
    path = 'path',
    rect = 'rect',
    animate = 'animate'
}

export enum SVG_PROPERTIES {
    viewBox = 'viewBox'
}

export const DECIMALS = 6;
export const SQRT3 = +(Math.sqrt(3).toFixed(DECIMALS));
export const HSQRT3 = +((Math.sqrt(3) / 2).toFixed(DECIMALS));
export const DEFAULT_WIDTH = 640;
export const DEFAULT_HEIGHT = 480;

const N = '(-?\\d+(?:\\.\\d+)?|-?\\.\\d+)'; // Number
const S = '\\s*'; // Space
const SS = '\\s+'; // Splicit space
const POINT = `${N}${SS}${N}${SS}${N}`;
const CURVE = `${POINT}${SS}${POINT}`;
export const COMMANDS_REGEXP = new RegExp(`(?:(M|L)${S}${POINT}${S}|(C)${S}${CURVE})${S}`, 'g');