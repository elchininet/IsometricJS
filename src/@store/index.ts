interface Sizes {
    centerX: number;
    centerY: number;
    height: number;
    width: number;
    scale: number;
}

export class Store {

    constructor(width: number, height: number, scale: number) {
        this.sizes = {
            centerX: width / 2,
            centerY: height / 2,
            height,
            width,
            scale
        };
    }

    private sizes: Sizes;
 
    public get width(): number {
        return this.sizes.width;
    }

    public set width(value: number) {
        this.sizes.width = value;
        this.sizes.centerX = value / 2;
    }

    public get height(): number {
        return this.sizes.height;
    }

    public set height(value: number) {
        this.sizes.height = value;
        this.sizes.centerY = value / 2;
    }

    public get scale(): number {
        return this.sizes.scale;
    }

    public set scale(value: number) {
        this.sizes.scale = value;
    }

    public get centerX(): number {
        return this.sizes.centerX;
    }

    public get centerY(): number {
        return this.sizes.centerY;
    }

}

