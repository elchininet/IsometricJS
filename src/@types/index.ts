export enum Colors {
    white = 'white',
    black = 'black'
}

export enum LineCap {
    butt = 'butt',
    square = 'square',
    round = 'round'
}

export enum LineJoin {
    miter = 'miter',
    round = 'round',
    bevel = 'bevel'
}

export type StrokeLinecap = keyof typeof LineCap;
export type StrokeLinejoin = keyof typeof LineJoin;

export interface Point {
    r?: number;
    l?: number;
    t?: number;
}

export interface Listener {
    fn: VoidFunction;
    fnBind: VoidFunction;
}

export enum Command {
    move = 'move',
    line = 'line',
    curve = 'curve'
}

export interface LinePoint {
    command: Exclude<Command, Command.curve>;
    control?: never;
    point: Point;
}

export interface CurvePoint {
    command: Exclude<Command, Command.move | Command.line>;
    control: Point;
    point: Point;
}

export type CommandPoint = LinePoint | CurvePoint;

export enum PlaneView {
    FRONT = 'FRONT',
    SIDE = 'SIDE',
    TOP = 'TOP'
}

export type IsometricPlaneView = keyof typeof PlaneView;