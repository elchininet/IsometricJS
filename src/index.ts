export { IsometricCanvas, IsometricCanvasProps } from '@classes/public/IsometricCanvas';
export { IsometricPath } from '@classes/public/IsometricPath';
export { GraphicProps as IsometricPathProps } from '@classes/abstract/Graphic';
export { IsometricRectangle, IsometricRectangleProps } from '@classes/public/IsometricRectangle';
export { IsometricCircle, IsometricCircleProps } from '@classes/public/IsometricCircle';
export { LineCap, LineJoin, PlaneView } from '@types';