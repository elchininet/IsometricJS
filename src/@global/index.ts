import { DEFAULT_WIDTH, DEFAULT_HEIGHT } from '@constants';

const store = {
    height: DEFAULT_HEIGHT,
    width: DEFAULT_WIDTH,
    centerX: DEFAULT_WIDTH / 2,
    centerY: DEFAULT_HEIGHT / 2,
    scale: 1
};

interface Sizes {
    height: number;
    width: number;
}

export class GlobalData {

    public static setSizes(width: number, height: number): void {
        GlobalData.height = height;
        GlobalData.width = width;
    }
 
    public static get width(): number {
        return store.width;
    }

    public static set width(value: number) {
        store.width = value;
        store.centerX = value / 2;
    }

    public static get height(): number {
        return store.height;
    }

    public static set height(value: number) {
        store.height = value;
        store.centerY = value / 2;
    }

    public static get scale(): number {
        return store.scale;
    }

    public static set scale(value: number) {
        store.scale = value;
    }

    public static get centerX(): number {
        return store.centerX;
    }

    public static get centerY(): number {
        return store.centerY;
    }

}

