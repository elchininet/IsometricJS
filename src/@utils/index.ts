import { Point, Listener, CommandPoint, Command } from '@types';
import { SQRT3, DECIMALS, COMMANDS_REGEXP } from '@constants';
import { IsometricPath } from '@classes/public/IsometricPath';

export interface IsometricPoint {
    x: number;
    y: number;
}

type IsometricControlPoints = [IsometricPoint, IsometricPoint, IsometricPoint, IsometricPoint];

export interface SVGProps {
    [key: string]: string;
}

const round = (n: number, d: number): number => {
    const exp = Math.pow(10, d);
    return Math.round(n * exp) / exp;
};

const getPointFromIsometricPoint = (
    centerX: number,
    centerY: number,
    point: Point,
    scale: number
): IsometricPoint => {
    return {
        x: round(centerX + (point.r - point.l) * scale * SQRT3, DECIMALS),
        y: round(centerY + ((point.r + point.l) / 2 - point.t) * scale, DECIMALS)
    };
};

const getControlPointsFromIsometricPoint = (
    centerX: number,
    centerY: number,
    fromPoint: Point,
    control: Point,
    toPoint: Point,
    scale: number
): IsometricControlPoints => {
    const corner1 = {
        r: control.r + (fromPoint.r - toPoint.r) / 2,
        l: control.l + (fromPoint.l - toPoint.l) / 2,
        t: control.t + (fromPoint.t - toPoint.t) / 2
    };
    const corner2 = {
        r: control.r - (fromPoint.r - toPoint.r) / 2,
        l: control.l - (fromPoint.l - toPoint.l) / 2,
        t: control.t - (fromPoint.t - toPoint.t) / 2
    };
    const isometricControlPoint1 = {
        r: corner1.r - (corner1.r - fromPoint.r) / 2,
        l: corner1.l - (corner1.l - fromPoint.l) / 2,
        t: corner1.t - (corner1.t - fromPoint.t) / 2
    };
    const isometricControlPoint2 = {
        r: control.r + (corner1.r - control.r) / 2,
        l: control.l + (corner1.l - control.l) / 2,
        t: control.t + (corner1.t - control.t) / 2
    };
    const isometricControlPoint3 = {
        r: control.r + (corner2.r - control.r) / 2,
        l: control.l + (corner2.l - control.l) / 2,
        t: control.t + (corner2.t - control.t) / 2
    };    
    const isometricControlPoint4 = {
        r: corner2.r - (corner2.r - toPoint.r) / 2,
        l: corner2.l - (corner2.l - toPoint.l) / 2,
        t: corner2.t - (corner2.t - toPoint.t) / 2
    };
    return [
        getPointFromIsometricPoint(centerX, centerY, isometricControlPoint1, scale),
        getPointFromIsometricPoint(centerX, centerY, isometricControlPoint2, scale),
        getPointFromIsometricPoint(centerX, centerY, isometricControlPoint3, scale),
        getPointFromIsometricPoint(centerX, centerY, isometricControlPoint4, scale)
    ];
};

export const addSVGProperties = (svg: SVGElement, props: SVGProps): void => {
    Object.keys(props).forEach((prop: string): void => {
        svg.setAttributeNS(null, prop, props[prop]);
    });
};

export const getSVGPath = (commands: CommandPoint[], centerX: number, centerY: number, scale: number): string => {
    const svgPaths = commands.map((c: CommandPoint, index: number) => {
        const point = getPointFromIsometricPoint(centerX, centerY, c.point, scale);     
        switch (c.command) {
            case Command.move:
                return `M${point.x} ${point.y}`;
            case Command.line:
                return `L${point.x} ${point.y}`;
            case Command.curve:
                if (index > 0) {
                    const middle = getPointFromIsometricPoint(centerX, centerY, c.control, scale);
                    const controls = getControlPointsFromIsometricPoint(
                        centerX,
                        centerY,
                        commands[index - 1].point,
                        c.control,
                        c.point,
                        scale
                    );
                    const middlePoint = `${middle.x} ${middle.y}`;
                    const finalPoint = `${point.x} ${point.y}`;
                    const control1 = `${controls[0].x} ${controls[0].y}`;
                    const control2 = `${controls[1].x} ${controls[1].y}`;
                    const control3 = `${controls[2].x} ${controls[2].y}`;
                    const control4 = `${controls[3].x} ${controls[3].y}`;
                    return `C${control1},${control2},${middlePoint} C${control3},${control4},${finalPoint}`;
                }
                return '';
            }
    });
    if (svgPaths.length) {
        const svgPathString = svgPaths.join(' ').trim();
        if (svgPathString) {
            return `${svgPathString}z`;
        }
    }
    return '';
};

export const drawCommands = (pathInstance: IsometricPath, commands: string): IsometricPath => {
    let array;
    while ((array = COMMANDS_REGEXP.exec(commands)) !== null) {
        const command = array[5] || array[1];
        switch(command) {
            case 'M':
                pathInstance.moveTo(+array[2], +array[3], +array[4]);
                break;
            case 'L':
                pathInstance.lineTo(+array[2], +array[3], +array[4]);
                break;
            case 'C':
                pathInstance.curveTo(+array[6], +array[7], +array[8], +array[9], +array[10], +array[11]);
        }
    }
    return pathInstance;
};

export const translateCommandPoints = (commands: CommandPoint[], right: number, left: number, top: number): void => {
    commands.forEach((command: CommandPoint): void => {
        command.point.r += right;
        command.point.l += left;
        command.point.t += top;
        // Comment for now because curves are not translated at the moment
        /*if (command.control) {
            command.control.r += right;
            command.control.l += left;
            command.control.t += top;
        }*/
    });
};

export function addEventListenerToElement(element: SVGElement, listeners: Listener[], event: string, callback: VoidFunction, useCapture: boolean): void {
    const listener = {
        fn: callback,
        fnBind: callback.bind(this)
    };
    listeners.push(listener);
    element.addEventListener(event, listener.fnBind, useCapture);
}

export function removeEventListenerFromElement(element: SVGElement, listeners: Listener[], event: string, callback: VoidFunction, useCapture: boolean): void {
    let listener: Listener;
    listeners.find((ln: Listener, index: number): boolean => {
        if (ln.fn === callback) {
            listener = listeners.splice(index, 1)[0];
            return true;
        }
    });
    if (listener) {
        element.removeEventListener(event, listener.fnBind, useCapture);
    }
}