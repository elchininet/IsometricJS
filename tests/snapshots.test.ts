import { IsometricCanvas, IsometricRectangle, IsometricCircle, IsometricPath, PlaneView } from '../src';

describe('Snapshot tests', (): void => {

    let container: HTMLDivElement;

    beforeEach((): void => {
        container = document.createElement('div');
        document.body.appendChild(container);
    });

    afterEach((): void => {
        if (container.parentNode && container.parentNode === document.body) {
            document.body.removeChild(container);
        }        
    });

    it('Draw rectangles', (): void => {

        const cube = new IsometricCanvas(container, {
            backgroundColor: '#CCC',
            scale: 120,
            width: 500,
            height: 320
        });

        const commonProps = {height: 1, width: 1};
        const topPiece = new IsometricRectangle({...commonProps, planeView: PlaneView.TOP});
        const rightPiece = new IsometricRectangle({...commonProps, planeView: PlaneView.FRONT});
        const leftPiece = new IsometricRectangle({...commonProps, planeView: PlaneView.SIDE});

        topPiece.top = 1;
        rightPiece.right = 1;
        leftPiece.left = 1;

        cube.addChild(topPiece).addChild(rightPiece).addChild(leftPiece);

        expect(container).toMatchSnapshot();

    });

    it('Draw circles', (): void => {

        const cube = new IsometricCanvas(container, {
            backgroundColor: '#CCC',
            scale: 120,
            width: 500,
            height: 320
        });

        const commonProps = {radius: 0.5};
        const topPiece = new IsometricCircle({...commonProps, planeView: PlaneView.TOP});
        const rightPiece = new IsometricCircle({...commonProps, planeView: PlaneView.FRONT});
        const leftPiece = new IsometricCircle({...commonProps, planeView: PlaneView.SIDE});

        topPiece.right = topPiece.left = 0.5;
        topPiece.top = 1;
        rightPiece.left = rightPiece.top = 0.5;
        rightPiece.right = 1;
        leftPiece.right = leftPiece.top = 0.5;
        leftPiece.left = 1;

        cube.addChild(topPiece).addChild(rightPiece).addChild(leftPiece);

        expect(container).toMatchSnapshot();

    });

    it('Draw methods', (): void => {

        const isometric = new IsometricCanvas(container, {
            backgroundColor: '#CCC',
            scale: 120,
            width: 500,
            height: 320
        });

        const bottomT = new IsometricPath();
        const bottomR = new IsometricPath();
        const bottomL = new IsometricPath();
        
        const topT = new IsometricPath();
        const topR = new IsometricPath();
        const topL = new IsometricPath();

        bottomT.moveTo(0, 0, .5).lineTo(1, 0, .5).lineTo(1, 1, .5).lineTo(0, 1, .5);
        bottomR.moveTo(1, 0, .5).lineTo(1, 0, 0).lineTo(1, 1, 0).lineTo(1, 1, .5);
        bottomL.moveTo(1, 1, .5).lineTo(1, 1, 0).lineTo(0, 1, 0).lineTo(0, 1, .5);

        topT.moveTo(.25, .25, 1.25).lineTo(.75, .25, 1.25).lineTo(.75, .75, 1).lineTo(.25, .75, 1);
        topR.moveTo(.75, .25, 1.25).lineTo(.75, .75, 1).lineTo(.75, .75, .5).lineTo(.75, .25, .5);
        topL.moveTo(.75, .75, 1).lineTo(.25, .75, 1).lineTo(.25, .75, .5).lineTo(.75, .75, .5);

        isometric.addChildren(bottomT, bottomR, bottomL, topT, topR, topL);

        expect(container).toMatchSnapshot();

    });

    it('Draw commands', (): void => {

        const isometric = new IsometricCanvas(container, {
            backgroundColor: '#CCC',
            scale: 120,
            width: 500,
            height: 320
        });

        const right = new IsometricPath();
        const top1 = new IsometricPath();
        const top2 = new IsometricPath();
        const top3 = new IsometricPath();
        const top4 = new IsometricPath();
        const left1 = new IsometricPath();
        const left2 = new IsometricPath();
        
        right.draw('M1 0 0 L1 1 0 L1 1 0.25 L1 0.5 0.25 L1 0.5 1 L1 0 1');
        top1.draw('M0.25 0.5 1 C0.5 0.5 0.75 0.75 0.5 1 L0.75 0 1 C0.5 0 0.75 0.25 0 1 L0.25 0.5 1');
        top2.draw('M1 0 1 L0.75 0 1 L0.75 0.5 1 L1 0.5 1 L1 0 1 M0 0 1 L0.25 0 1 L0.25 0.5 1 L0 0.5 1 L0 0 1');
        top3.draw('M0 0.5 0.5 L0.5 0.5 0.5 L0.5 1 0.5 L0 1 0.5');
        top4.draw('M0.5 0.5 0.5 L1 0.5 0.25 L1 1 0.25 L0.5 1 0.5');
        left1.draw('M0 0.5 1 L0 0.5 0.5 L0.5 0.5 0.5 L1 0.5 0.25 L1 0.5 1 L0.75 0.5 1 C0.5 0.5 0.75 0.25 0.5 1 L0 0.5 1');
        left2.draw('M0 1 0.5 L0.5 1 0.5 L1 1 0.25 L1 1 0 L0 1 0');

        isometric.addChildren(right, top1, top2, top3, top4, left1, left2);

        expect(container).toMatchSnapshot();

    });

    it('Draw curves with method aliases', (): void => {

        const cube = new IsometricCanvas(container, {
            backgroundColor: '#CCC',
            scale: 120,
            width: 500,
            height: 320
        });
    
        const under = new IsometricPath({ fillColor: '#EEE' });
        const top = new IsometricPath();
        const right = new IsometricPath();
        const left = new IsometricPath();
    
        under
            .mt(0, 0, 1)
            .mt(0.25, 0, 1).ct(0.5, 0, 0.75, 0.75, 0, 1).lt(1, 0, 1)
            .lt(1, 0, 0.75).ct(0.75, 0, 0.5, 1, 0, 0.25).lt(1, 0, 0)
            .lt(1, 0.25, 0).ct(0.75, 0.5, 0, 1, 0.75, 0).lt(1, 1, 0)
            .lt(0.75, 1, 0).ct(0.5, 0.75, 0, 0.25, 1, 0).lt(0, 1, 0)
            .lt(0, 1, 0.25).ct(0, 0.75, 0.5, 0, 1, 0.75).lt(0, 1, 1)
            .lt(0, 0.75, 1).ct(0, 0.5, 0.75, 0, 0.25, 1).lt(0, 0, 1);
    
        top
            .mt(0, 0, 1)
            .lt(0.25, 0, 1).ct(0.5, 0.25, 1, 0.75, 0, 1).lt(1, 0, 1)
            .lt(1, 0.25, 1).ct(0.75, 0.5, 1, 1, 0.75, 1).lt(1, 1, 1)        
            .lt(0.75, 1, 1).ct(0.5, 0.75, 1, 0.25, 1, 1).lt(0, 1, 1)
            .lt(0, 0.75, 1).ct(0.25, 0.5, 1, 0, 0.25, 1).lt(0, 0, 1);
    
        right
            .mt(1, 0, 1)
            .lt(1, 0, 0.75).ct(1, 0.25, 0.5, 1, 0, 0.25).lt(1, 0, 0)
            .lt(1, 0.25, 0).ct(1, 0.5, 0.25, 1, 0.75, 0).lt(1, 1, 0)
            .lt(1, 1, 0.25).ct(1, 0.75, 0.5, 1, 1, 0.75).lt(1, 1, 1)
            .lt(1, 0.75, 1).ct(1, 0.5, 0.75, 1, 0.25, 1).lt(1, 0, 1);
    
        left
            .mt(1, 1, 1)
            .lt(1, 1, 0.75).ct(0.75, 1, 0.5, 1, 1, 0.25).lt(1, 1, 0)
            .lt(0.75, 1, 0).ct(0.5, 1, 0.25, 0.25, 1, 0).lt(0, 1, 0)
            .lt(0, 1, 0.25).ct(0.25, 1, 0.5, 0, 1, 0.75).lt(0, 1, 1)
            .lt(0.25, 1, 1).ct(0.5, 1, 0.75, 0.75, 1, 1).lt(1, 1, 1);
    
        cube.addChildren(under, top, right, left);

        expect(container).toMatchSnapshot();

    });



});