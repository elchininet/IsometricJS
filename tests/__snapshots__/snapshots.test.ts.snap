// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot tests Draw circles 1`] = `
<div>
  <svg
    height="320px"
    viewBox="0 0 500 320"
    width="500px"
  >
    <rect
      fill="#CCC"
      height="320px"
      width="500px"
      x="0"
      y="0"
    />
    <path
      d="M198.0385 130 A 73.484658 42.426407 0 0 0 301.9615 70 A 73.484658 42.426407 180 0 0 198.0385 130z"
      fill="white"
      fill-opacity="1"
      stroke="black"
      stroke-dasharray=""
      stroke-linecap="butt"
      stroke-linejoin="round"
      stroke-opacity="1"
      stroke-width="1"
    />
    <path
      d="M250 220 A 73.484684 42.426392 119.999977 0 0 353.923 160 A 73.484684 42.426392 -60.000023 0 0 250 220z"
      fill="white"
      fill-opacity="1"
      stroke="black"
      stroke-dasharray=""
      stroke-linecap="butt"
      stroke-linejoin="round"
      stroke-opacity="1"
      stroke-width="1"
    />
    <path
      d="M146.077 160 A 73.484684 42.426392 60.000023 0 0 250 220 A 73.484684 42.426392 -119.999977 0 0 146.077 160z"
      fill="white"
      fill-opacity="1"
      stroke="black"
      stroke-dasharray=""
      stroke-linecap="butt"
      stroke-linejoin="round"
      stroke-opacity="1"
      stroke-width="1"
    />
  </svg>
</div>
`;

exports[`Snapshot tests Draw commands 1`] = `
<div>
  <svg
    height="320px"
    viewBox="0 0 500 320"
    width="500px"
  >
    <rect
      fill="#CCC"
      height="320px"
      width="500px"
      x="0"
      y="0"
    />
    <path
      d="M353.923 220 L250 280 L250 250 L301.9615 220 L301.9615 130 L353.923 100z"
      fill="white"
      fill-opacity="1"
      stroke="black"
      stroke-dasharray=""
      stroke-linecap="butt"
      stroke-linejoin="round"
      stroke-opacity="1"
      stroke-width="1"
    />
    <path
      d="M224.01925 85 A 36.742342 21.213196 60.000023 0 0 275.98075 115 L327.94225 85 A 36.742342 21.213196 60.000023 0 1 275.98075 55 L224.01925 85z"
      fill="white"
      fill-opacity="1"
      stroke="black"
      stroke-dasharray=""
      stroke-linecap="butt"
      stroke-linejoin="round"
      stroke-opacity="1"
      stroke-width="1"
    />
    <path
      d="M353.923 100 L327.94225 85 L275.98075 115 L301.9615 130 L353.923 100 M250 40 L275.98075 55 L224.01925 85 L198.0385 70 L250 40z"
      fill="white"
      fill-opacity="1"
      stroke="black"
      stroke-dasharray=""
      stroke-linecap="butt"
      stroke-linejoin="round"
      stroke-opacity="1"
      stroke-width="1"
    />
    <path
      d="M198.0385 130 L250 160 L198.0385 190 L146.077 160z"
      fill="white"
      fill-opacity="1"
      stroke="black"
      stroke-dasharray=""
      stroke-linecap="butt"
      stroke-linejoin="round"
      stroke-opacity="1"
      stroke-width="1"
    />
    <path
      d="M250 160 L301.9615 220 L250 250 L198.0385 190z"
      fill="white"
      fill-opacity="1"
      stroke="black"
      stroke-dasharray=""
      stroke-linecap="butt"
      stroke-linejoin="round"
      stroke-opacity="1"
      stroke-width="1"
    />
    <path
      d="M198.0385 70 L198.0385 130 L250 160 L301.9615 220 L301.9615 130 L275.98075 115 A 36.742342 21.213196 60.000023 0 1 224.01925 85 L198.0385 70z"
      fill="white"
      fill-opacity="1"
      stroke="black"
      stroke-dasharray=""
      stroke-linecap="butt"
      stroke-linejoin="round"
      stroke-opacity="1"
      stroke-width="1"
    />
    <path
      d="M146.077 160 L198.0385 190 L250 250 L250 280 L146.077 220z"
      fill="white"
      fill-opacity="1"
      stroke="black"
      stroke-dasharray=""
      stroke-linecap="butt"
      stroke-linejoin="round"
      stroke-opacity="1"
      stroke-width="1"
    />
  </svg>
</div>
`;

exports[`Snapshot tests Draw curves with method aliases 1`] = `
<div>
  <svg
    height="320px"
    viewBox="0 0 500 320"
    width="500px"
  >
    <rect
      fill="#CCC"
      height="320px"
      width="500px"
      x="0"
      y="0"
    />
    <path
      d="M250 40 M275.98075 55 A 36.742342 21.213196 60.000023 0 0 327.94225 85 L353.923 100 L353.923 130 A 36.742342 21.213196 -119.999977 0 0 353.923 190 L353.923 220 L327.94225 235 A 36.742329 21.213203 -180 0 0 275.98075 265 L250 280 L224.01925 265 A 36.742329 21.213203 0 0 0 172.05775 235 L146.077 220 L146.077 190 A 36.742342 21.213196 -60.000023 0 0 146.077 130 L146.077 100 L172.05775 85 A 36.742342 21.213196 119.999977 0 0 224.01925 55 L250 40z"
      fill="#EEE"
      fill-opacity="1"
      stroke="black"
      stroke-dasharray=""
      stroke-linecap="butt"
      stroke-linejoin="round"
      stroke-opacity="1"
      stroke-width="1"
    />
    <path
      d="M250 40 L275.98075 55 A 36.742329 21.213203 180 0 0 327.94225 85 L353.923 100 L327.94225 115 A 36.742329 21.213203 -180 0 0 275.98075 145 L250 160 L224.01925 145 A 36.742329 21.213203 0 0 0 172.05775 115 L146.077 100 L172.05775 85 A 36.742329 21.213203 0 0 0 224.01925 55 L250 40z"
      fill="white"
      fill-opacity="1"
      stroke="black"
      stroke-dasharray=""
      stroke-linecap="butt"
      stroke-linejoin="round"
      stroke-opacity="1"
      stroke-width="1"
    />
    <path
      d="M353.923 100 L353.923 130 A 36.742342 21.213196 119.999977 0 0 353.923 190 L353.923 220 L327.94225 235 A 36.742342 21.213196 -60.000023 0 0 275.98075 265 L250 280 L250 250 A 36.742342 21.213196 -60.000023 0 0 250 190 L250 160 L275.98075 145 A 36.742342 21.213196 119.999977 0 0 327.94225 115 L353.923 100z"
      fill="white"
      fill-opacity="1"
      stroke="black"
      stroke-dasharray=""
      stroke-linecap="butt"
      stroke-linejoin="round"
      stroke-opacity="1"
      stroke-width="1"
    />
    <path
      d="M250 160 L250 190 A 36.742342 21.213196 -119.999977 0 0 250 250 L250 280 L224.01925 265 A 36.742342 21.213196 -119.999977 0 0 172.05775 235 L146.077 220 L146.077 190 A 36.742342 21.213196 60.000023 0 0 146.077 130 L146.077 100 L172.05775 115 A 36.742342 21.213196 60.000023 0 0 224.01925 145 L250 160z"
      fill="white"
      fill-opacity="1"
      stroke="black"
      stroke-dasharray=""
      stroke-linecap="butt"
      stroke-linejoin="round"
      stroke-opacity="1"
      stroke-width="1"
    />
  </svg>
</div>
`;

exports[`Snapshot tests Draw methods 1`] = `
<div>
  <svg
    height="320px"
    viewBox="0 0 500 320"
    width="500px"
  >
    <rect
      fill="#CCC"
      height="320px"
      width="500px"
      x="0"
      y="0"
    />
    <path
      d="M250 100 L353.923 160 L250 220 L146.077 160z"
      fill="white"
      fill-opacity="1"
      stroke="black"
      stroke-dasharray=""
      stroke-linecap="butt"
      stroke-linejoin="round"
      stroke-opacity="1"
      stroke-width="1"
    />
    <path
      d="M353.923 160 L353.923 220 L250 280 L250 220z"
      fill="white"
      fill-opacity="1"
      stroke="black"
      stroke-dasharray=""
      stroke-linecap="butt"
      stroke-linejoin="round"
      stroke-opacity="1"
      stroke-width="1"
    />
    <path
      d="M250 220 L250 280 L146.077 220 L146.077 160z"
      fill="white"
      fill-opacity="1"
      stroke="black"
      stroke-dasharray=""
      stroke-linecap="butt"
      stroke-linejoin="round"
      stroke-opacity="1"
      stroke-width="1"
    />
    <path
      d="M250 40 L301.9615 70 L250 130 L198.0385 100z"
      fill="white"
      fill-opacity="1"
      stroke="black"
      stroke-dasharray=""
      stroke-linecap="butt"
      stroke-linejoin="round"
      stroke-opacity="1"
      stroke-width="1"
    />
    <path
      d="M301.9615 70 L250 130 L250 190 L301.9615 160z"
      fill="white"
      fill-opacity="1"
      stroke="black"
      stroke-dasharray=""
      stroke-linecap="butt"
      stroke-linejoin="round"
      stroke-opacity="1"
      stroke-width="1"
    />
    <path
      d="M250 130 L198.0385 100 L198.0385 160 L250 190z"
      fill="white"
      fill-opacity="1"
      stroke="black"
      stroke-dasharray=""
      stroke-linecap="butt"
      stroke-linejoin="round"
      stroke-opacity="1"
      stroke-width="1"
    />
  </svg>
</div>
`;

exports[`Snapshot tests Draw rectangles 1`] = `
<div>
  <svg
    height="320px"
    viewBox="0 0 500 320"
    width="500px"
  >
    <rect
      fill="#CCC"
      height="320px"
      width="500px"
      x="0"
      y="0"
    />
    <path
      d="M250 40 L353.923 100 L250 160 L146.077 100z"
      fill="white"
      fill-opacity="1"
      stroke="black"
      stroke-dasharray=""
      stroke-linecap="butt"
      stroke-linejoin="round"
      stroke-opacity="1"
      stroke-width="1"
    />
    <path
      d="M353.923 220 L250 280 L250 160 L353.923 100z"
      fill="white"
      fill-opacity="1"
      stroke="black"
      stroke-dasharray=""
      stroke-linecap="butt"
      stroke-linejoin="round"
      stroke-opacity="1"
      stroke-width="1"
    />
    <path
      d="M146.077 220 L250 280 L250 160 L146.077 100z"
      fill="white"
      fill-opacity="1"
      stroke="black"
      stroke-dasharray=""
      stroke-linecap="butt"
      stroke-linejoin="round"
      stroke-opacity="1"
      stroke-width="1"
    />
  </svg>
</div>
`;
